docs/page-objects/additional-requests.rst

```diff
@@ -773,3 +773,53 @@ Here's an example:
 From the example above, we're now checking the list of responses to see if any
 exceptions are included in it. If so, we're simply logging it down and ignoring
 it. In this way, perfectly good responses can still be processed through.
+
+
+.. _retries-additional-requests:
+
+Retrying Additional Requests
+============================
+
+When the bad response data comes from :ref:`additional requests
+<additional-requests>`, you must handle retries on your own.
+
+The page object code is responsible for retrying additional requests until good
+response data is received, or until some maximum number of retries is exceeded.
+
+It is up to you to decide what the maximum number of retries should be for a
+given additional request, based on your experience with the target website.
+
+It is also up to you to decide how to implement retries of additional requests.
+
+One option would be tenacity_. For example, to try an additional request 3
+times before giving up:
+
+.. _tenacity: https://tenacity.readthedocs.io/en/latest/index.html
+
+.. code-block:: python
+
+    import attrs
+    from tenacity import retry, stop_after_attempt
+    from web_poet import HttpClient, WebPage
+
+    @attrs.define
+    class MyPage(WebPage):
+        http: HttpClient
+
+        @retry(stop=stop_after_attempt(3))
+        async def get_data(self):
+            response = await self.http.get("https://toscrape.com/")
+            if not response.css(".expected"):
+                raise ValueError
+            return response.css(".data").get()
+
+        async def to_item(self) -> dict:
+            try:
+                data = await self.get_data()
+            except ValueError:
+                return {}
+            return {"data": data}
```

Comment:
That's an example of a page object where validate_input won't work properly, if user decides to extend the page object with some validation. User might need to do something like this:

```suggestion
        async def to_item(self) -> dict:        
            if item := self._validate_input():
                return item
            try:
                data = await self.get_data()
            except ValueError:
                return {}
            return {"data": data}
```

which doesn't look very nice, and uses a private method as well.

One may also try `item = super().to_item()`, e.g. to add more information in a subclass. But it could be wrong. For example, validate_input of this class (or of a base class) might return an item. We say that this item is going to be used instead of to_item result. And indeed, `super().to_item()` is going to return this item, even if validate_input is implemented in a subclass. But then to_item method may start adding more things to the item, which means that the item is going to be different.

Reply:
We can definitely make `_validate_input` public. Would that be enough?

Alternatively, we could reconsider switching the “override item” logic from a return value to another exception, so that here we can go from 2 lines to 1 (i.e. save the return line).

Reply:
[Done](https://github.com/scrapinghub/web-poet/pull/113/commits/9819d57049b33cc7014cf8cb21bd4027a8c75d29).

The decorator is used in ItemPage.to_item, so existing tests cover the async scenario.

For the sync scenario, we had `test_invalid_input_sync_to_item` already, but it was passing. It was passing because the custom `to_item` implementation happened to be returning the same object as it would with input validation properly enabled, because input validation was still applied at the field level. I added an extra `is_valid` field to the `Item` class in tests that defaults to `True`, so that it is only `False` when that field is set in an `invalid_input` method. After that, the test started failing until you added the `@validate_input` decorator.